cli/crowdinApi.ts

```diff
@@ -0,0 +1,371 @@
+}
+
+async function getAllDirectories() {
+    // This request takes a decent amount of time, so cache the results
+    if (!fetchedDirectories) {
+        const { sourceFilesApi } = getClient();
+
+        pxt.debug(`Fetching directories`)
+        const dirsResponse = await sourceFilesApi
+            .withFetchAll()
+            .listProjectDirectories(projectId, {});
+
+        let dirs = dirsResponse.data.map(fileResponse => fileResponse.data);
+
+        if (!dirs.length) {
+            throw new Error("No directories found!");
+        }
+
+        pxt.debug(`Directory count: ${dirs.length}`);
+
+        fetchedDirectories = dirs;
+    }
+
+    return fetchedDirectories;
+}
+
+async function getAllFiles() {
+    // This request takes a decent amount of time, so cache the results
+    if (!fetchedFiles) {
+        const { sourceFilesApi } = getClient();
+
+        pxt.debug(`Fetching files`)
+        const filesResponse = await sourceFilesApi
+            .withFetchAll()
+            .listProjectFiles(projectId, {});
+
+        let files = filesResponse.data.map(fileResponse => fileResponse.data);
+
+        if (!files.length) {
+            throw new Error("No files found!");
+        }
+
+        pxt.debug(`File count: ${files.length}`);
+        fetchedFiles = files;
+    }
+
+    return fetchedFiles;
+}
+
+async function createFile(fileName: string, fileContent: any, directoryId?: number): Promise<void> {
+    const { uploadStorageApi, sourceFilesApi } = getClient();
+
+    // This request happens in two parts: first we upload the file to the storage API,
+    // then we actually create the file
+    const storageResponse = await uploadStorageApi.addStorage(fileName, fileContent);
+
+    const file = await sourceFilesApi.createFile(projectId, {
+        storageId: storageResponse.data.id,
+        name: fileName,
+        directoryId
+    });
+
+    // Make sure to add the file to the cache if it exists
+    if (fetchedFiles) {
+        fetchedFiles.push(file.data);
+    }
+}
+
+async function createDirectory(dirName: string, directoryId?: number): Promise<SourceFilesModel.Directory> {
+    const { sourceFilesApi } = getClient();
+
+    const dir = await sourceFilesApi.createDirectory(projectId, {
+        name: dirName,
+        directoryId
+    });
+
+    // Make sure to add the directory to the cache if it exists
+    if (fetchedDirectories) {
+        fetchedDirectories.push(dir.data);
+    }
+    return dir.data;
+}
+
+
+async function updateFile(fileId: number, fileName: string, fileContent: any): Promise<void> {
+    const { uploadStorageApi, sourceFilesApi } = getClient();
+
+    const storageResponse = await uploadStorageApi.addStorage(fileName, fileContent);
+
+    await sourceFilesApi.updateOrRestoreFile(projectId, fileId, {
+        storageId: storageResponse.data.id,
+    });
+}
+
+function getClient() {
+    if (!client) {
+        client = new crowdin(crowdinCredentials());
```

Comment:
might want to pass opts with a retryConfig as there have historically been a decent number of spurious errors (pass that in pxt-translations along with ignoring 304's and it was very necessary at the time at least)